core/src/test/kotlin/fr/sncf/osrd/stdcm/STDCMPathfindingTests.kt

```diff
@@ -672,4 +672,77 @@ class STDCMPathfindingTests {
                 .run()!!
         assertEquals(departureTime, res.departureTime)
     }
+
+    /**
+     * Train goes through start window without added delay, but before planned arrival time: check
+     * that 10s are added to departure time to make train pass in planned step at planned arrival =
+     * 10s.
+     */
+    @Test
+    fun testDelayAddedWhenTrainPassesBeforePlannedStepArrivalTime() {
+        val infra = DummyInfra()
+        val blocks = listOf(infra.addBlock("a", "b"))
+        val plannedStepArrivalTime = 10.0
+        val res =
+            STDCMPathfindingBuilder()
+                .setInfra(infra.fullInfra())
+                .setStartLocations(
+                    setOf(EdgeLocation(blocks[0], Offset(0.meters))),
+                    PlannedTimingData(plannedStepArrivalTime.seconds, 10.seconds, 10.seconds)
+                )
+                .setEndLocations(setOf(EdgeLocation(blocks[0], Offset(100.meters))))
+                .run()!!
+        assertEquals(plannedStepArrivalTime, res.departureTime)
+    }
+
+    /**
+     * Train goes through start window without added delay, before planned arrival time, but an
+     * occupancy starts at 5s, with an internal margin of 2s: check that 5s - 2s are added to
+     * departure time to make train pass in planned step closer to planned arrival = 10s.
+     */
+    @Test
+    fun testMaxPossibleDelayAddedWhenTrainPassesBeforePlannedStepArrivalTime() {
+        val infra = DummyInfra()
+        val blocks = listOf(infra.addBlock("a", "b"))
+        val plannedStepArrivalTime = 10.0
+        val startOccupancy = 5.0
+        val internalMargin = 2.0
+        val maxPossibleDelay = startOccupancy - internalMargin
+        val occupancyGraph =
+            ImmutableMultimap.of(
+                blocks[0],
+                OccupancySegment(startOccupancy, 10_000.0, 0.meters, 0.meters)
+            )
+        val res =
+            STDCMPathfindingBuilder()
+                .setInfra(infra.fullInfra())
+                .setStartLocations(
+                    setOf(EdgeLocation(blocks[0], Offset(0.meters))),
+                    PlannedTimingData(plannedStepArrivalTime.seconds, 10.seconds, 10.seconds)
+                )
+                .setEndLocations(setOf(EdgeLocation(blocks[0], Offset(100.meters))))
+                .setUnavailableTimes(occupancyGraph)
+                .run()!!
+        assertEquals(maxPossibleDelay, res.departureTime)
+    }
+
+    /**
+     * Train goes through start window without added delay, after planned arrival time: no delay
```

Comment:
We could add a test with a lot of added delay right from the start

Reply:
Don't quite get this.
Is this a new test? If so, what would it be testing?
Or just the same test, because you'd rather have an occupancy at the start rather than a hack negative planned arrival time? (honestly thought it was fine, we have a bit of leeway with test set up in my opinion).

Reply:
In fact I have a specific test case in mind, I could write it

Reply:
Oh you mean check that we actually add the edge delay and the planned delay. Yeah ok I see the test. It's fine i can prolly write it ;)